dali/operators/reader/loader/video/frames_decoder_gpu.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_READER_LOADER_VIDEO_FRAMES_DECODER_GPU_H_
+#define DALI_OPERATORS_READER_LOADER_VIDEO_FRAMES_DECODER_GPU_H_
+
+#include "dali/operators/reader/loader/video/frames_decoder.h"
+
+#include <string>
+#include <memory>
+#include <queue>
+#include <vector>
+
+#include "dali/operators/reader/loader/video/nvdecode/cuviddec.h"
+#include "dali/operators/reader/loader/video/nvdecode/nvcuvid.h"
+
+#include "dali/core/dev_buffer.h"
+
+namespace dali {
+struct NvDecodeState {
+  CUvideodecoder decoder_;
+  CUvideoparser parser_;
+
+  CUVIDSOURCEDATAPACKET packet = { 0 };
+
+  uint8_t *decoded_frame_yuv;
+};
+
+struct BufferedFrame {
+  DeviceBuffer<uint8_t> frame_;
+  int pts_;
+};
+
+class DLL_PUBLIC FramesDecoderGpu : public FramesDecoder {
+ public:
+  /**
+   * @brief Construct a new FramesDecoder object.
+   * 
+   * @param filename Path to a video file.
+   */
+  explicit FramesDecoderGpu(const std::string &filename, cudaStream_t stream = 0);
+
+  bool ReadNextFrame(uint8_t *data, bool copy_to_output = true) override;
+
+  void SeekFrame(int frame_id) override;
+
+  void Reset() override;
+
+  int CurrentFramePts() { return index_[CurrentFrame()].pts; }
+
+  std::unique_ptr<NvDecodeState> nvdecode_state_;
```

Comment:
```suggestion
private:
  std::unique_ptr<NvDecodeState> nvdecode_state_;
```

Reply:
Can't do it, because I need to access it from the callback.

Alternative would be to make the callback to call a public method that accesses it but it seemed a bit convoluted. `FramesDecoderGpu` is not supposed to be used by the outside users, so I think maybe it is fine for now.

Reply:
How about making the callback to call the public method that corresponds to the callback?:
```
int process_picture_decode(void *user_data, CUVIDPICPARAMS *picture_params) {
  FramesDecoderGpu *frames_decoder = static_cast<FramesDecoderGpu*>(user_data);
  return frames_decoder->process_picture_decode_impl(picture_params);
}
```
I would still opt for making the data private.

Reply:
Done